examples/android/routeguide/app/src/main/java/io/grpc/routeguideexample/RouteGuideActivity.java

```diff
@@ -0,0 +1,426 @@
+        mRecordRouteButton = (Button) findViewById(R.id.record_route_button);
+        mRouteChatButton = (Button) findViewById(R.id.route_chat_button);
+        mResultText = (TextView) findViewById(R.id.result_text);
+        mResultText.setMovementMethod(new ScrollingMovementMethod());
+        disableButtons();
+    }
+    
+    public void startRouteGuide(View view) {
+        String host = mHostEdit.getText().toString();
+        String portStr = mPortEdit.getText().toString();
+        int port = TextUtils.isEmpty(portStr) ? 0 : Integer.valueOf(portStr);
+        ((InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE))
+                .hideSoftInputFromWindow(mHostEdit.getWindowToken(), 0);
+        mChannel = ManagedChannelBuilder.forAddress(host, port).usePlaintext(true).build();
+        mHostEdit.setEnabled(false);
+        mPortEdit.setEnabled(false);
+        mStartRouteGuideButton.setEnabled(false);
+        enableButtons();
+    }
+    
+    public void exitRouteGuide(View view) {
+        mChannel.shutdown();
+        disableButtons();
+        mHostEdit.setEnabled(true);
+        mPortEdit.setEnabled(true);
+        mStartRouteGuideButton.setEnabled(true);
+    }
+
+    public void getFeature(View view) {
+        new GrpcTask(new GetFeatureRunnable()).execute();
+    }
+
+    public void listFeatures(View view) {
+        new GrpcTask(new ListFeaturesRunnable()).execute();
+    }
+
+    public void recordRoute(View view) {
+        new GrpcTask(new RecordRouteRunnable()).execute();
+    }
+
+    public void routeChat(View view) {
+        new GrpcTask(new RouteChatRunnable()).execute();
+    }
+    
+    private void disableButtons() {
+        mGetFeatureButton.setEnabled(false);
+        mListFeaturesButton.setEnabled(false);
+        mRecordRouteButton.setEnabled(false);
+        mRouteChatButton.setEnabled(false);
+        mExitRouteGuideButton.setEnabled(false);
+    }
+    
+    private void enableButtons() {
+        mExitRouteGuideButton.setEnabled(true);
+        mGetFeatureButton.setEnabled(true);
+        mListFeaturesButton.setEnabled(true);
+        mRecordRouteButton.setEnabled(true);
+        mRouteChatButton.setEnabled(true);
+    }
+
+    private class GrpcTask extends AsyncTask<Void, Void, String> {
+        private final GrpcRunnable mGrpc;
+
+        GrpcTask(GrpcRunnable grpc) {
+            this.mGrpc = grpc;
+        }
+
+        @Override
+        protected void onPreExecute() {
+            mResultText.setText("");
+            disableButtons();
+        }
+
+        @Override
+        protected String doInBackground(Void... nothing) {
+            try {
+                String logs = mGrpc.run(RouteGuideGrpc.newBlockingStub(mChannel),
+                        RouteGuideGrpc.newStub(mChannel));
+                return "Success!" + System.lineSeparator() + logs;
+            } catch (Exception e) {
+                return "Failed... : " + System.lineSeparator() + e.getMessage();
```

Comment:
Many of our exceptions require looking at the cause to figure out what's going on (including basic stuff, like wrong hostname). Please return the full stack trace